solr/core/src/test/org/apache/solr/search/TestMainQueryCaching.java

```diff
@@ -0,0 +1,313 @@
+    assertEquals("Bad full sort count", expectFullSortCount, coreToSortCount(core, "full"));
+    assertEquals("Bad skip sort count", expectSkipSortCount, coreToSortCount(core, "skip"));
+    assertEquals("Should have exactly " + expectNumFound, expectNumFound, (long) (body.get("numFound"))); // sanity check
+  }
+
+  @Test
+  public void testConcurrentMatchAllDocsInitialization() throws Exception {
+    final int nThreads = 20;
+    final ExecutorService executor = ExecutorUtil.newMDCAwareFixedThreadPool(nThreads, new SolrNamedThreadFactory(getTestName()));
+    final Future<?>[] followup = new Future<?>[nThreads];
+    for (int i = 0; i < nThreads; i++) {
+      final int myI = i;
+      followup[i] = executor.submit(() -> {
+        try {
+          // NOTE: we use cursorMark=* here because it prevents consulting the queryResultCache, which can interfere
+          // with DocSet fetching (which is what we care about in this test).
+          String response = JQ(req("q", MATCH_ALL_DOCS_QUERY, "request_id", Integer.toString(myI), "cursorMark", "*", "sort", "id asc"));
+          Map<?, ?> res = (Map<?, ?>) fromJSONString(response);
+          Map<?, ?> body = (Map<?, ?>) (res.get("response"));
+          assertEquals("Should have exactly " + ALL_DOCS, ALL_DOCS, (long) (body.get("numFound"))); // sanity check
+        } catch (Exception ex) {
+          throw new RuntimeException(ex);
+        }
+      });
+    }
+    try {
+      for (Future<?> f : followup) {
+        f.get(); // to access exceptions/errors
+      }
+    } finally {
+      executor.shutdown();
+      assertTrue(executor.awaitTermination(5, TimeUnit.SECONDS)); // tasks should already have completed
+    }
+    final SolrCore core = h.getCore();
+    Map<String, Object> liveDocsCacheMetrics = coreToLiveDocsCacheMetrics(core);
+    long inserts = (long) liveDocsCacheMetrics.get("inserts"); // the one and only liveDocs computation
+    long hits = (long) liveDocsCacheMetrics.get("hits"); // hits during the initial phase
+    long asyncHits = (long) liveDocsCacheMetrics.get("asyncHits");
+    long naiveHits = coreToLiveDocsNaiveCacheHitCount(core);
+
+    assertEquals(1, inserts);
+    assertEquals(nThreads - 1, hits + naiveHits);
+    assertTrue(asyncHits <= hits);
+
+    // NOTE: The assertion below is commented out because, although it may _often_ be true, it is dependent
+    // on timing/thread scheduling; in practice it happens that not infrequently `asyncHits == 0` (e.g., if matchAllDocs
+    // computation happens quickly, and/or if subsequent threads were delayed).
+    //
+    // It seems that the assertion below more frequently succeeds when this test is run in isolation; e.g.:
+    // `gradlew :solr:core:test --tests "org.apache.solr.search.TestMainQueryCaching.testConcurrentMatchAllDocsInitialization"`
+
+    //assertTrue("expected asyncHits > 0; found asyncHits=" + asyncHits, asyncHits > 0);
```

Comment:
We could introduce delay at some point to cause threads to stall via TestInjection, but I'm fine without it.

Reply:
That'd be really cool. I actually kind of knocked myself out a little trying to do something like that but didn't want to reinvent the wheel or end up with something that needlessly slows down the test suite. If you know of a good (dare I say "right"?) way to do this I'm definitely interested.